MinRTT for the path (#2919)
quicwg/base-drafts: draft-ietf-quic-recovery.md

At a high level, an endpoint measures the time from when a packet was sent to
when it is acknowledged as a round-trip time (RTT) sample.  The endpoint uses
RTT samples and peer-reported host delays (see Section 13.2 of

> {{QUIC-TRANSPORT}}) to generate a statistical description of the path's RTT.

Reviewer: ```suggestion
{{QUIC-TRANSPORT}}) to generate a statistical description of the network path's RTT.
```

## Estimating min_rtt {#min-rtt}

> min_rtt is the minimum RTT observed for a given path.  min_rtt is set to the

Reviewer: ```suggestion
min_rtt is the minimum RTT observed for a given network path.  min_rtt is set to the
```

latest_rtt on the first RTT sample, and to the lesser of min_rtt and latest_rtt
on subsequent samples.  In this document, min_rtt is used by loss detection to
reject implausibly small rtt samples.

An endpoint uses only locally observed times in computing the min_rtt and does
not adjust for host delays reported by the peer.  Doing so allows the endpoint
to set a lower bound for the smoothed_rtt based entirely on what it observes
(see {{smoothed-rtt}}), and limits potential underestimation due to
erroneously-reported delays by the peer.

> The RTT for a path may change over time.  If a path's actual RTT decreases,

Reviewer: ```suggestion
The RTT for a network path may change over time.  If a path's actual RTT decreases,
```

> the min_rtt will adapt immediately on the first low sample.  If the path's

Reviewer: ```suggestion
the min_rtt will adapt immediately on the first low sample.  If the network path's
```

At a high level, an endpoint measures the time from when a packet was sent to
when it is acknowledged as a round-trip time (RTT) sample.  The endpoint uses
RTT samples and peer-reported host delays ({{host-delay}}) to generate a
statistical description of the path's RTT.  An endpoint computes the
following three values: the minimum value observed over the lifetime of the

> congestion control context (min_rtt), an exponentially-weighted moving average

Reviewer: ```suggestion
RTT estimator's context (min_rtt), an exponentially-weighted moving average
```

Author: We never define this, but isn't the RTT estimator a subset of the congestion control context?  Also, we only use RTT estimator once in this document at the moment.

Possibly I should use path for all of these now, against @erickinnear suggestion?  Also, path is much shorter than "congestion control context"

  min_rtt.  This limits the underestimation that a misreporting peer can cause
  to the smoothed_rtt.

> On the first RTT sample for a path, the smoothed_rtt is set to the latest_rtt.

Reviewer: ```suggestion
On the first RTT sample for a network path, the smoothed_rtt is set to the latest_rtt.
```

smoothed_rtt and rttvar are computed as follows, similar to {{?RFC6298}}.  On

> the first RTT sample for a path:

Reviewer: ```suggestion
the first RTT sample for a network path:
```

## Estimating min_rtt {#min-rtt}

> min_rtt is the minimum RTT observed for a given congestion control context.

Reviewer: min_rtt is only used for loss recovery in this doc. I think it's more appropriate to say RTT estimator context in general.

Reviewer 2: When you say "congestion control context", why not "path"?

Author: Good question, I'll talk with Eric and Jana in person about this one and we can figure out some consistent terminology.

min_rtt is set to the latest_rtt on the first RTT sample, and to the lesser
of min_rtt and latest_rtt on subsequent samples.  min_rtt is used to reject

> implausible RTT samples, but is not used directly in congestion control or

Reviewer: I would drop the rest of the text after "samples" and perhaps add "which would otherwise cause the smoothed_rtt estimate to be incorrectly low.". I understand what you're trying to say, but I think saying that it is not used directly in loss recovery makes it a bit confusing I think. It is used, and I think people will have different reads on what's indirect and what's

loss recovery.

An endpoint uses only locally observed times in computing the min_rtt and does
not adjust for host delays reported by the peer ({{host-delay}}).  Doing so
allows the endpoint to set a lower bound for the smoothed_rtt based entirely on
what it observes (see {{smoothed-rtt}}), and limits potential underestimation
due to erroneously-reported delays by the peer.

> The RTT for the path may change over time.  If the real path RTT decreases,

Reviewer: ```suggestion
The RTT for a path may change over time.  If a path's actual RTT decreases,
```

> it should be observed quickly.  If the real path RTT increases, it will not

Reviewer: ```suggestion
the min_rtt will adapt immediately on the first low sample.  If the path's actual RTT increases,
```

> be observed, but the impact is limited to allowing some RTT samples that are

Reviewer: ```suggestion
the min_rtt will not adapt to it, potentially causing future RTT samples that are
```

  min_rtt.  This limits the underestimation that a misreporting peer can cause
  to the smoothed_rtt.

> On the first RTT sample in a congestion control context, the smoothed_rtt is

Reviewer: RTT context

Author: path?

Reviewer 3: Elsewhere don’t we discuss resetting a congestion control context? It seems like that often lines up with a path, but they don’t have to be the same thing. It seems like you’re saying that RTT isn’t part of congestion control (which sounds totally valid but we then need some better way to word that)?

set to the latest_rtt.

smoothed_rtt and rttvar are computed as follows, similar to {{?RFC6298}}.  On

> the first RTT sample in a congestion control context:

Reviewer: RTT context

Author: path?
